neqo-qpack/src/encoder.rs

```diff
@@ -257,28 +271,44 @@ impl QPackEncoder {
         }
     }
 
-    fn change_capacity(&mut self, value: u64) -> Res<()> {
+    fn need_change_capacity(&mut self, value: u64) {
         qdebug!([self], "change capacity: {}", value);
-        self.table.set_capacity(value)?;
-        EncoderInstruction::Capacity { value }.marshal(&mut self.send_buf, self.use_huffman);
+        self.change_capacity = Some(value);
+    }
+
+    fn maybe_send_change_capacity(&mut self, conn: &mut Connection, stream_id: u64) -> Res<()> {
+        if let Some(cap) = self.change_capacity.take() {
+            let mut buf = QPData::default();
+            EncoderInstruction::Capacity { value: cap }.marshal(&mut buf, self.use_huffman);
+            if !conn.stream_send_atomic(stream_id, &buf)? {
+                return Err(Error::EncoderStreamBlocked);
+            }
+            self.table.set_capacity(cap)?;
```

Comment:
As we have two fallible operations here, this is tricky.  I would suggest that you try to this (and attempt to perform evictions) first as that operation - if it succeeds - won't result in a half-committed state.  If we manage to set capacity, but not send the instruction, then we can still try to set the capacity again later before writing another instruction.  If we can't evict, but send the instruction, then we will have violated the protocol (probably).